ospf6d/ospf6_bfd.c

```diff
@@ -44,72 +44,41 @@
- */
-void ospf6_bfd_info_free(void **bfd_info)
-{
-	bfd_info_free((struct bfd_info **)bfd_info);
-}
-
-/*
- * ospf6_bfd_show_info - Show BFD info structure
- */
-void ospf6_bfd_show_info(struct vty *vty, void *bfd_info, int param_only,
-			 json_object *json_obj, bool use_json)
-{
-	if (param_only)
-		bfd_show_param(vty, bfd_info, 1, 0, use_json, json_obj);
-	else
-		bfd_show_info(vty, bfd_info, 0, 1, use_json, json_obj);
-}
-
-/*
- * ospf6_bfd_reg_dereg_nbr - Register/Deregister a neighbor with BFD through
- *                           zebra for starting/stopping the monitoring of
- *                           the neighbor rechahability.
- */
-void ospf6_bfd_reg_dereg_nbr(struct ospf6_neighbor *on, int command)
-{
-	struct ospf6_interface *oi = on->ospf6_if;
-	struct interface *ifp = oi->interface;
-	struct bfd_info *bfd_info;
-	char src[64];
-	int cbit;
-
-	if (!oi->bfd_info || !on->bfd_info)
-		return;
-	bfd_info = (struct bfd_info *)oi->bfd_info;
-
-	if (IS_OSPF6_DEBUG_ZEBRA(SEND)) {
-		inet_ntop(AF_INET6, &on->linklocal_addr, src, sizeof(src));
-		zlog_debug("%s nbr (%s) with BFD",
-			   bfd_get_command_dbg_str(command), src);
-	}
-
-	cbit = CHECK_FLAG(bfd_info->flags, BFD_FLAG_BFD_CBIT_ON);
-
-	bfd_peer_sendmsg(zclient, bfd_info, AF_INET6, &on->linklocal_addr,
-			 on->ospf6_if->linklocal_addr, ifp->name, 0, 0, cbit,
-			 command, 0, ifp->vrf_id);
-
-	if (command == ZEBRA_BFD_DEST_DEREGISTER)
-		bfd_info_free((struct bfd_info **)&on->bfd_info);
-}
-
 /*
  * ospf6_bfd_trigger_event - Neighbor is registered/deregistered with BFD when
  *                           neighbor state is changed to/from 2way.
  */
 void ospf6_bfd_trigger_event(struct ospf6_neighbor *on, int old_state,
 			     int state)
 {
-	if ((old_state < OSPF6_NEIGHBOR_TWOWAY)
-	    && (state >= OSPF6_NEIGHBOR_TWOWAY))
-		ospf6_bfd_reg_dereg_nbr(on, ZEBRA_BFD_DEST_REGISTER);
-	else if ((old_state >= OSPF6_NEIGHBOR_TWOWAY)
-		 && (state < OSPF6_NEIGHBOR_TWOWAY))
-		ospf6_bfd_reg_dereg_nbr(on, ZEBRA_BFD_DEST_DEREGISTER);
+	int family;
+	struct in6_addr src, dst;
+
+	/* Skip sessions without BFD. */
+	if (on->bfd_session == NULL)
+		return;
+
+	if (old_state < OSPF6_NEIGHBOR_TWOWAY
+	    && state >= OSPF6_NEIGHBOR_TWOWAY) {
+		/*
+		 * Check if neighbor address changed.
+		 *
+		 * When the neighbor is configured BFD before having an existing
+		 * connection, then the destination address will be set to `::`
+		 * which will cause session installation failure. This piece of
+		 * code updates the address in that case.
+		 */
+		bfd_sess_addresses(on->bfd_session, &family, &src, &dst);
+		if (memcmp(&on->linklocal_addr, &dst, sizeof(dst))) {
+			bfd_sess_set_ipv6_addrs(on->bfd_session, &src,
+						&on->linklocal_addr);
+		}
+
+		bfd_sess_enable(on->bfd_session, true);
+		bfd_sess_install(on->bfd_session);
+	} else if (old_state >= OSPF6_NEIGHBOR_TWOWAY
+		   && state < OSPF6_NEIGHBOR_TWOWAY)
+		bfd_sess_enable(on->bfd_session, false);
```

Comment:
Same question. Should be `bfd_sess_uninstall`.

Reply:
Fixed